### ISSUE TITLE
Read BioArgo data

### ISSUE THREAD

**Issue by CaraWilson** [dankelley/oce#812]
I'd like to use OCE routines with BioArgo data that is available on MBARI's website at http://www.mbari.org/chemsensor/floatviz.htm

These files aren't in the standard ARGO format so can't be read by read.drifter.  I've read the data into R (example code below) but am unable to convert it into CTD class (since drifter class doesn’t allow for additional parameters) using as.ctd since the data is only provided as depth, and not pressure as required.  

The code below reads in one example file from the MBARI site. It would need to be more flexible to read in all of the available files since different files has different number of parameters depending on which sensors are on the different floats, and the headers are different sizes.
   
```r
BAurl <- url("http://www.mbari.org/lobo/Data/FloatVizData/QC/8486HawaiiQC.txt")

# Header has repeating variable names (each parameter has a "QF", so 
# header=T can't be used by read.table).  Additionally each profile has 
# a row without any values, so have to use Fill=T with read.table.
# These rows are removed are the file is read in. 

cnames=c("Float","Station","Type","Date","Time","Long","Lat","Bdepth","Depth","Depth_QF",
         "Temp","Temp_QF","Salinity","Sal_QF","Dens","Dens_QF","Oxygen","Oxygen_QF",
         "OxygenSat","OxygenSat_QF","Nitrate","Nitrate_QF","Chlorophyll","Chlorophyll_QF",
         "Backscatter", "Backscatter_QF","pHinsitu","pHinsitu_QF","pH25C","ph25C_QF")
colClasses <- rep("numeric",30)
colClasses[1:5] <- "character"
no_col <- max(count.fields(BAurl, sep = "\t"))

float <- read.table(BAurl,sep="\t",fill=TRUE,skip=24,col.names=cnames,colClasses=colClasses)
float$Date <- paste(float$Date,float$Time)
float$Date <- parse_date_time(float$Date, "%m%d%Y %H%M")
float$Station<-as.numeric(float$Station)

# Remove rows where Depth = NA 

float<-subset(float, !is.na(Depth))
```

**Comment by richardsc** [comment]
Hi @CaraWilson ,

Thanks for the request. I edited your comment a bit to clean up the code section by wrapping it in backticks to indicate a verbatim chunk (just click the "pencil" button in the upper right of your post to see what I mean).

I'm not too familiar with the BioARGO floats (yet), but I expect that we'll be able to figure out a way to read the files into sensible `oce` objects.

Are you set up to build packages from source? I ask because the most recent `develop` branch of oce has some differences from the current CRAN version (we are planning a new release in the next few weeks), including an object name change from `drifter` to `argo`. We are also starting to think about quality flags (e.g. the `_QF` fields in your example), so this might a good test for some implementation.

If you're not set up to build packages, perhaps you can check out the instructions at:

https://github.com/dankelley/oce/wiki/Configuring-a-system-to-be-able-to-build-oce-from-source

Otherwise, @dankelley and I will have a look, and will document and as you questions here as the issue progresses.

**Comment by richardsc** [comment]
Should have mentioned that once you're set up to build packages, you can install the latest version from Github by following the instructions under "Source-code development version (advanced users)" here:

http://dankelley.github.io/oce/installation.html

**Comment by dankelley** [comment]
Hi. I can see basically how to handle this, and can do that tomorrow morning. I wonder, though -- is there a URL that describes the format? It would be smart to handle not just a particular file, but general files. For example, if there are specially-coded comments that we can parse, it would help to know the rules for their coding. Similarly, if the system permits different commenting characters, we should handle them all. 

The basic code that I'll write will do this:
1. read a bunch of lines to try to figure out where the header ends
2. go back one or several lines, to try to infer column titles
3. read the lines after the header
Of these, item 2 is generally the trickiest, and that's where a statement about the format is helpful. For example, If both "Temperature" and "T" can occur, we'd want to know that. Also, we'd want to know the units (e.g. temperature scale), etc.

**One more thing...** um, the data file you sent does not match the format that the URL http://www.mbari.org/chemsensor/floatviz.htm provides. So which do you want to read? (Sorry, just checked the url after typing this.)

**Comment by CaraWilson** [comment]
Hi Dan - 

As far as I can tell there is nothing on the website that describes the file formats.  I did look at a few different files and it seems the names are consistent between different floats that have different sensors on them, i.e. Temperature is always Temperature (in the few files I looked at).  And the header row does contain the units for the parameters, for example: 

Cruise	Station	Type	mon/day/yr	hh:mm	Lon [°E]	Lat [°N]	Bot. Depth [m]	Depth[m]	QF	Temperature[°C]	QF	Salinity	QF	Density	QF	Oxygen[µM]	QF	OxygenSat[%]	QF	Nitrate[µM]	QF

I didn’t read in the header because when I tried doing that it didn’t like the repeating “QF” flags. There is probably a way to deal with that but I’m not very proficient at R.

I know Ken Johnson, whose email is listed on one of the pages to send comments to, and if you think it would help I’d be happy to send him an email asking him about this. 

Cara 


> On Jan 4, 2016, at 11:03 AM, Dan Kelley <notifications@github.com> wrote:
> 
> Hi. I can see basically how to handle this, and can do that tomorrow morning. I wonder, though -- is there a URL that describes the format? It would be smart to handle not just a particular file, but general files. For example, if there are specially-coded comments that we can parse, it would help to know the rules for their coding. Similarly, if the system permits different commenting characters, we should handle them all.
> 
> The basic code that I'll write will do this:
> 1. read a bunch of lines to try to figure out where the header ends
> 2. go back one or several lines, to try to infer column titles
> 3. read the lines after the header
> Of these, item 2 is generally the trickiest, and that's where a statement about the format is helpful. For example, If both "Temperature" and "T" can occur, we'd want to know that. Also, we'd want to know the units (e.g. temperature scale), etc.
> 
> One more thing... um, the data file you sent does not match the format that the URL http://www.mbari.org/chemsensor/floatviz.htm <http://www.mbari.org/chemsensor/floatviz.htm> provides. So which do you want to read? (Sorry, just checked the url after typing this.)
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/dankelley/oce/issues/812#issuecomment-168771672>.
> 


______________________________

Cara Wilson
NOAA/NMFS/SWFSC                                      
Environmental Research Division 
Southwest Fisheries Science Center 
99 Pacific Street, Suite 255A
Monterey, CA 93940

Phone:  831-648-5337
Fax:      831-648-8440
cara.wilson@noaa.gov
http://www.pfeg.noaa.gov/~cwilson

**Comment by dankelley** [comment]
Maybe you could ask that person. It would help, fo example, to know whether longitude will always be `Lon` or whether it may sometimes be `Longitude`. Another issue relates to the units. In the example you show, I think there is a space before the `[` for longitude, but not for nitrate. These are the sorts of things that it's nice to know beforehand.

PS. it helps a lot if you use markdown to highlight text, e.g. put three backticks on a line before that `Cruise` line, etc. I tried doing that but it did not "take", perhaps because you entered in rich text or something.

**Comment by dankelley** [comment]
Sorry, one more thing and then I go offline for the day... how exactly will you get the data? Will you click the `SEND` button and then cut/paste into a file? This matters because then we get into issues like the encoding on your computer (given that we have a Greek 'mu' letter there) and sometimes cut/paste will change tabs into spaces, etc. Parsing that header will be a ton easier if the items are tab-separated or comma-separated; otherwise it is really quite tricky because we won't know beforehand what might be there, and I guess in what order. The method of parsing the header will likely be to use `strsplit()`, which is a handy function if the data are in a simple format. If there are no tab characters in the file, then parsing is going to be tricky, e.g. we cannot use space to separate because some items have multiple words, and we can't use upper-case letters because times don't employ that convention.

I do think it would help to point Ken Jackson at this website so he can see my comments. I'm offline now for the day but perhaps Clark will have time during his day to contribute more.

**Comment by richardsc** [comment]
Just piping in from my digging around, it looks like there are two ways to get the data from that site, each providing differently formatted files.

1. The first way involves choosing the desired float from http://www.mbari.org/chemsensor/FloatViz.htm and then clicking `Text file` and the `SEND` button. Note that to get multiple data columns this way you have to Ctrl-click the "Y variables".

2. The second is to click the [Float list and link to complete Ascii data files](http://www.mbari.org/chemsensor/FloatList.html) and then just choose the data file from the table. It looks like the files @CaraWilson are interested in are the ones obtained this way.

My suggestion would be to code only for the second option, but I agree it is worth an email to Ken to ask about the data formats, etc. I met him at the GAIC meeting in Galway last fall, and I think he'd love to know that we're trying to figure out how to make the data more accessible.

A relevant question might be if there is a alternate source for the BioArgo data -- e.g. are there NetCDF files contained on the Argo DACs?

**Comment by richardsc** [comment]
> A relevant question might be if there is a alternate source for the BioArgo data -- e.g. are there NetCDF files contained on the Argo DACs?

Going to answer my own question a bit, after some digging. It looks like most of the floats listed at:

http://www.mbari.org/chemsensor/FloatList.html

have WMO numbers (the number in brackets, like 5904124 for the example float `8486HawaiiQC`), which means the data are stored in the various Argo DACs. For example, putting that float id into the search box at

http://www.argodatamgt.org/Access-to-data/Description-of-all-floats2

brings up a page which provides a link to the NetCDF file:

http://www.ifremer.fr/co-diffClient/diffusion?ptfCode=5904124&format=Netcdf&type=profiles&lang=en&mode=Argo

(note that it takes a bit of time to generate the file before downloading).

The next question will be: "Does oce read BioArgo files appropriately?". I'll try and see once I get the NetCDF from the above link.

**Comment by CaraWilson** [comment]
Hi Clark and Dan - 

I will shoot an email to Ken and ask him about this, I will cc you two on it as well. Thanks do much for your help with this! I was intending to get the data the second way the Clark mentioned - i.e. the direct link to the txt file as I coded in the example I sent you. 

Cara 

PS I’m rather new to github - would it be preferable for me to respond you your emails there?  


> On Jan 4, 2016, at 12:06 PM, Clark Richards <notifications@github.com> wrote:
> 
> Just piping in from my digging around, it looks like there are two ways to get the data from that site, each providing differently formatted files.
> 
> The first way involves choosing the desired float from http://www.mbari.org/chemsensor/FloatViz.htm <http://www.mbari.org/chemsensor/FloatViz.htm> and then clicking Text file and the SEND button. Note that to get multiple data columns this way you have to Ctrl-click the "Y variables".
> 
> The second is to click the Float list and link to complete Ascii data files <http://www.mbari.org/chemsensor/FloatList.html> and then just choose the data file from the table. It looks like the files @CaraWilson <https://github.com/CaraWilson> are interested in are the ones obtained this way.
> 
> My suggestion would be to code only for the second option, but I agree it is worth an email to Ken to ask about the data formats, etc. I met him at the GAIC meeting in Galway last fall, and I think he'd love to know that we're trying to figure out how to make the data more accessible.
> 
> A relevant question might be if there is a alternate source for the BioArgo data -- e.g. are there NetCDF files contained on the Argo DACs?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/dankelley/oce/issues/812#issuecomment-168788917>.
> 


______________________________

Cara Wilson
NOAA/NMFS/SWFSC                                      
Environmental Research Division 
Southwest Fisheries Science Center 
99 Pacific Street, Suite 255A
Monterey, CA 93940

Phone:  831-648-5337
Fax:      831-648-8440
cara.wilson@noaa.gov
http://www.pfeg.noaa.gov/~cwilson

**Comment by richardsc** [comment]
Hi Cara,

Responding to the Github emails is fine if that's easiest for you, but it can be good if you strip out any of the "quoted" text, as it tends to cause the Issue discussion to become "bloated" with extra text. I've been stripping them out of your comments as they come in, which isn't a big deal either.

I still haven't managed to download the NetCDF file from the Ifremer server -- it keeps timing out on me. I might have to try one of the other DACs if it doesn't work. I'll update here, but it would be helpful to know what the difference is between the files provided through MBARI and the "official" Argo files.

Clark

**Comment by CaraWilson** [comment]
I just sent Ken an email letting him know about this discussion, so hopefully he will provide some feedback on the issue. I’d also like to know the difference between the data on the MBARI website and the official ARGO files. 

Cara 


> On Jan 4, 2016, at 12:19 PM, Clark Richards <notifications@github.com> wrote:
> 
> Hi Cara,
> 
> Responding to the Github emails is fine if that's easiest for you, but it can be good if you strip out any of the "quoted" text, as it tends to cause the Issue discussion to become "bloated" with extra text. I've been stripping them out of your comments as they come in, which isn't a big deal either.
> 
> I still haven't managed to download the NetCDF file from the Ifremer server -- it keeps timing out on me. I might have to try one of the other DACs if it doesn't work. I'll update here, but it would be helpful to know what the difference is between the files provided through MBARI and the "official" Argo files.
> 
> Clark
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/dankelley/oce/issues/812#issuecomment-168792791>.
> 


______________________________

Cara Wilson
NOAA/NMFS/SWFSC                                      
Environmental Research Division 
Southwest Fisheries Science Center 
99 Pacific Street, Suite 255A
Monterey, CA 93940

Phone:  831-648-5337
Fax:      831-648-8440
cara.wilson@noaa.gov
http://www.pfeg.noaa.gov/~cwilson

**Comment by richardsc** [comment]
Ok, I got the file. If you want to skip the wait, you can download it from:

https://www.dropbox.com/s/aulcu4lkgbg205d/5904124_20160104211204622.nc?dl=0

A quick try suggests that oce can read it as an `argo`/`drifter` object, but it's only reading the "standard" argo fields:
```r
library(oce)
d <- read.oce('5904124_20160104211204622.nc')
str(d)
```
```
Formal class 'argo' [package "oce"] with 3 slots
  ..@ metadata     :List of 3
  .. ..$ filename: chr "/Users/richardsc/Dropbox (Personal)/misc/Rstuff/bioargo/5904124_20160104211204622.nc"
  .. ..$ units   : list()
  .. ..$ id      : chr "5904124"
  ..@ data         :List of 6
  .. ..$ time       : POSIXct[1:273], format: "2013-05-20 14:33:37" "2013-05-21 17:48:09" ...
  .. ..$ longitude  : num [1:273(1d)] -158 -158 -158 -158 -158 ...
  .. ..$ latitude   : num [1:273(1d)] 22.7 22.7 22.7 22.7 22.7 ...
  .. ..$ salinity   : num [1:496, 1:273] 34.8 34.8 34.8 34.8 34.8 ...
  .. ..$ temperature: num [1:496, 1:273] 25.1 25.1 25.1 25.1 25.1 ...
  .. ..$ pressure   : num [1:496, 1:273] 7.7 11.8 16.4 21.7 26.7 ...
  ..@ processingLog:List of 2
  .. ..$ time : POSIXct[1:2], format: "2016-01-04 15:19:22" "2016-01-04 15:19:22"
  .. ..$ value: chr [1:2] "create 'argo' object" "read.argo(file = file)"
```

To verify that all the "bio" data are there, we can load the NetCDF file directly and look at the contents:
```r
library(ncdf4)
nc <- nc_open('5904124_20160104211204622.nc')
nc
```
```
File 5904124_20160104211204622.nc (NC_FORMAT_CLASSIC):

     76 variables (excluding dimension variables):
        char DATA_TYPE[STRING16]   
            long_name: Data type
            _FillValue:  
        char FORMAT_VERSION[STRING4]   
            long_name: File format version
            _FillValue:  
        char HANDBOOK_VERSION[STRING4]   
            long_name: Data handbook version
            _FillValue:  
        char REFERENCE_DATE_TIME[DATE_TIME]   
            long_name: Date of reference for Julian days
            conventions: YYYYMMDDHHMISS
            _FillValue:  
        char DATE_CREATION[DATE_TIME]   
            long_name: Date of file creation
            conventions: YYYYMMDDHHMISS
            _FillValue:  
        char DATE_UPDATE[DATE_TIME]   
            long_name: Date of update of this file
            conventions: YYYYMMDDHHMISS
            _FillValue:  
        char PLATFORM_NUMBER[STRING8,N_PROF]   
            long_name: Float unique identifier
            conventions: WMO float identifier : A9IIIII
            _FillValue:  
        char PROJECT_NAME[STRING64,N_PROF]   
            long_name: Name of the project
            _FillValue:  
        char PI_NAME[STRING64,N_PROF]   
            long_name: Name of the principal investigator
            _FillValue:  
        char STATION_PARAMETERS[STRING16,N_PARAM,N_PROF]   
            long_name: List of available parameters for the station
            conventions: Argo reference table 3
            _FillValue:  
        int CYCLE_NUMBER[N_PROF]   
            long_name: Float cycle number
            conventions: 0..N, 0 : launch cycle (if exists), 1 : first complete cycle
            _FillValue: 99999
        char DIRECTION[N_PROF]   
            long_name: Direction of the station profiles
            conventions: A: ascending profiles, D: descending profiles
            _FillValue:  
        char DATA_CENTRE[STRING2,N_PROF]   
            long_name: Data centre in charge of float data processing
            conventions: Argo reference table 4
            _FillValue:  
        char DC_REFERENCE[STRING32,N_PROF]   
            long_name: Station unique identifier in data centre
            conventions: Data centre convention
            _FillValue:  
        char DATA_STATE_INDICATOR[STRING4,N_PROF]   
            long_name: Degree of processing the data have passed through
            conventions: Argo reference table 6
            _FillValue:  
        char DATA_MODE[N_PROF]   
            long_name: Delayed mode or real time data
            conventions: R : real time; D : delayed mode; A : real time with adjustment
            _FillValue:  
        char PLATFORM_TYPE[STRING32,N_PROF]   
            long_name: Type of float
            _FillValue:  
        char FLOAT_SERIAL_NO[STRING16,N_PROF]   
            long_name: Serial number of the float
            _FillValue:  
        char FIRMWARE_VERSION[STRING16,N_PROF]   
            long_name: Instrument version
            _FillValue:  
        char WMO_INST_TYPE[STRING4,N_PROF]   
            long_name: Coded instrument type
            conventions: Argo reference table 8
            _FillValue:  
        double JULD[N_PROF]   
            long_name: Julian day (UTC) of the station relative to REFERENCE_DATE_TIME
            standard_name: time
            units: days since 1950-01-01 00:00:00 UTC
            conventions: Relative julian days with decimal part (as parts of day)
            _FillValue: 999999
            axis: T
        char JULD_QC[N_PROF]   
            long_name: Quality on Date and Time
            conventions: Argo reference table 2
            _FillValue:  
        double JULD_LOCATION[N_PROF]   
            long_name: Julian day (UTC) of the location relative to REFERENCE_DATE_TIME
            units: days since 1950-01-01 00:00:00 UTC
            conventions: Relative julian days with decimal part (as parts of day)
            _FillValue: 999999
        double LATITUDE[N_PROF]   
            long_name: Latitude of the station, best estimate
            standard_name: latitude
            units: degree_north
            _FillValue: 99999
            valid_min: -90
            valid_max: 90
            axis: Y
        double LONGITUDE[N_PROF]   
            long_name: Longitude of the station, best estimate
            standard_name: longitude
            units: degree_east
            _FillValue: 99999
            valid_min: -180
            valid_max: 180
            axis: X
        char POSITION_QC[N_PROF]   
            long_name: Quality on position (latitude and longitude)
            conventions: Argo reference table 2
            _FillValue:  
        char POSITIONING_SYSTEM[STRING8,N_PROF]   
            long_name: Positioning system
            _FillValue:  
        char PROFILE_PRES_QC[N_PROF]   
            long_name: Global quality flag of PRES profile
            conventions: Argo reference table 2a
            _FillValue:  
        char PROFILE_PSAL_QC[N_PROF]   
            long_name: Global quality flag of PSAL profile
            conventions: Argo reference table 2a
            _FillValue:  
        char PROFILE_DOXY_QC[N_PROF]   
            long_name: Global quality flag of DOXY profile
            conventions: Argo reference table 2a
            _FillValue:  
        char PROFILE_TEMP_QC[N_PROF]   
            long_name: Global quality flag of TEMP profile
            conventions: Argo reference table 2a
            _FillValue:  
        char PROFILE_TEMP_DOXY_QC[N_PROF]   
            long_name: Global quality flag of TEMP_DOXY profile
            conventions: Argo reference table 2a
            _FillValue:  
        char VERTICAL_SAMPLING_SCHEME[STRING256,N_PROF]   
            long_name: Vertical sampling scheme
            conventions: Argo reference table 16
            _FillValue:  
        int CONFIG_MISSION_NUMBER[N_PROF]   
            long_name: Float's mission number for each profile
            conventions: 0..N, 0 : launch mission (if exists), 1 : first complete mission
            _FillValue: 99999
        float PRES[N_LEVELS,N_PROF]   
            long_name: SEA PRESSURE
            standard_name: sea_water_pressure
            _FillValue: 99999
            units: decibar
            valid_min: 0
            valid_max: 12000
            C_format: %7.1f
            FORTRAN_format: F7.1
            resolution: 0.100000001490116
            axis: Z
        char PRES_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float PRES_ADJUSTED[N_LEVELS,N_PROF]   
            long_name: SEA PRESSURE
            standard_name: sea_water_pressure
            _FillValue: 99999
            units: decibar
            valid_min: 0
            valid_max: 12000
            C_format: %7.1f
            FORTRAN_format: F7.1
            resolution: 0.100000001490116
            axis: Z
        char PRES_ADJUSTED_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float PRES_ADJUSTED_ERROR[N_LEVELS,N_PROF]   
            long_name: SEA PRESSURE
            _FillValue: 99999
            units: decibar
            C_format: %7.1f
            FORTRAN_format: F7.1
            resolution: 0.100000001490116
        float PSAL[N_LEVELS,N_PROF]   
            long_name: PRACTICAL SALINITY
            standard_name: sea_water_salinity
            _FillValue: 99999
            units: psu
            valid_min: 0
            valid_max: 42
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char PSAL_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float PSAL_ADJUSTED[N_LEVELS,N_PROF]   
            long_name: PRACTICAL SALINITY
            standard_name: sea_water_salinity
            _FillValue: 99999
            units: psu
            valid_min: 0
            valid_max: 42
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char PSAL_ADJUSTED_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float PSAL_ADJUSTED_ERROR[N_LEVELS,N_PROF]   
            long_name: PRACTICAL SALINITY
            _FillValue: 99999
            units: psu
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        float DOXY[N_LEVELS,N_PROF]   
            long_name: DISSOLVED OXYGEN
            standard_name: moles_of_oxygen_per_unit_mass_in_sea_water
            _FillValue: 99999
            units: micromole/kg
            valid_min: 0
            valid_max: 650
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char DOXY_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float DOXY_ADJUSTED[N_LEVELS,N_PROF]   
            long_name: DISSOLVED OXYGEN
            standard_name: moles_of_oxygen_per_unit_mass_in_sea_water
            _FillValue: 99999
            units: micromole/kg
            valid_min: 0
            valid_max: 650
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char DOXY_ADJUSTED_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float DOXY_ADJUSTED_ERROR[N_LEVELS,N_PROF]   
            long_name: DISSOLVED OXYGEN
            _FillValue: 99999
            units: micromole/kg
            valid_min: 0
            valid_max: 650
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        float TEMP[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE IN SITU ITS-90 SCALE
            standard_name: sea_water_temperature
            _FillValue: 99999
            units: degree_Celsius
            valid_min: -2
            valid_max: 40
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char TEMP_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float TEMP_ADJUSTED[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE IN SITU ITS-90 SCALE
            standard_name: sea_water_temperature
            _FillValue: 99999
            units: degree_Celsius
            valid_min: -2
            valid_max: 40
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char TEMP_ADJUSTED_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float TEMP_ADJUSTED_ERROR[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE IN SITU ITS-90 SCALE
            _FillValue: 99999
            units: degree_Celsius
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        float TEMP_DOXY[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE FROM DOXY SENSOR (ITS-90 SCALE)
            standard_name: temperature_of_sensor_for_oxygen_in_sea_water
            _FillValue: 99999
            units: degree_Celsius
            valid_min: -2
            valid_max: 40
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char TEMP_DOXY_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float TEMP_DOXY_ADJUSTED[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE FROM DOXY SENSOR (ITS-90 SCALE)
            standard_name: temperature_of_sensor_for_oxygen_in_sea_water
            _FillValue: 99999
            units: degree_Celsius
            valid_min: -2
            valid_max: 40
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char TEMP_DOXY_ADJUSTED_QC[N_LEVELS,N_PROF]   
            long_name: quality flag
            conventions: Argo reference table 2
            _FillValue:  
        float TEMP_DOXY_ADJUSTED_ERROR[N_LEVELS,N_PROF]   
            long_name: SEA TEMPERATURE FROM DOXY SENSOR (ITS-90 SCALE)
            _FillValue: 99999
            units: degree_Celsius
            valid_min: -2
            valid_max: 40
            C_format: %9.3f
            FORTRAN_format: F9.3
            resolution: 0.00100000004749745
        char PARAMETER[STRING16,N_PARAM,N_CALIB,N_PROF]   
            long_name: List of parameters with calibration information
            conventions: Argo reference table 3
            _FillValue:  
        char SCIENTIFIC_CALIB_EQUATION[STRING256,N_PARAM,N_CALIB,N_PROF]   
            long_name: Calibration equation for this parameter
            _FillValue:  
        char SCIENTIFIC_CALIB_COEFFICIENT[STRING256,N_PARAM,N_CALIB,N_PROF]   
            long_name: Calibration coefficients for this equation
            _FillValue:  
        char SCIENTIFIC_CALIB_COMMENT[STRING256,N_PARAM,N_CALIB,N_PROF]   
            long_name: Comment applying to this parameter calibration
            _FillValue:  
        char SCIENTIFIC_CALIB_DATE[DATE_TIME,N_PARAM,N_CALIB,N_PROF]   
            long_name: Date of calibration
            _FillValue:  
        char HISTORY_INSTITUTION[STRING4,N_PROF,N_HISTORY]   
            long_name: Institution which performed action
            conventions: Argo reference table 4
            _FillValue:  
        char HISTORY_STEP[STRING4,N_PROF,N_HISTORY]   
            long_name: Step in data processing
            conventions: Argo reference table 12
            _FillValue:  
        char HISTORY_SOFTWARE[STRING4,N_PROF,N_HISTORY]   
            long_name: Name of software which performed action
            conventions: Institution dependent
            _FillValue:  
        char HISTORY_SOFTWARE_RELEASE[STRING4,N_PROF,N_HISTORY]   
            long_name: Version/release of software which performed action
            conventions: Institution dependent
            _FillValue:  
        char HISTORY_REFERENCE[STRING64,N_PROF,N_HISTORY]   
            long_name: Reference of database
            conventions: Institution dependent
            _FillValue:  
        char HISTORY_DATE[DATE_TIME,N_PROF,N_HISTORY]   
            long_name: Date the history record was created
            conventions: YYYYMMDDHHMISS
            _FillValue:  
        char HISTORY_ACTION[STRING4,N_PROF,N_HISTORY]   
            long_name: Action performed on data
            conventions: Argo reference table 7
            _FillValue:  
        char HISTORY_PARAMETER[STRING16,N_PROF,N_HISTORY]   
            long_name: Station parameter action is performed on
            conventions: Argo reference table 3
            _FillValue:  
        float HISTORY_START_PRES[N_PROF,N_HISTORY]   
            long_name: Start pressure action applied on
            _FillValue: 99999
            units: decibar
        float HISTORY_STOP_PRES[N_PROF,N_HISTORY]   
            long_name: Stop pressure action applied on
            _FillValue: 99999
            units: decibar
        float HISTORY_PREVIOUS_VALUE[N_PROF,N_HISTORY]   
            long_name: Parameter/Flag previous value before action
            _FillValue: 99999
        char HISTORY_QCTEST[STRING16,N_PROF,N_HISTORY]   
            long_name: Documentation of tests performed, tests failed (in hex form)
            conventions: Write tests performed when ACTION=QCP$; tests failed when ACTION=QCF$
            _FillValue:  

     13 dimensions:
        DATE_TIME  Size:14
        STRING256  Size:256
        STRING64  Size:64
        STRING32  Size:32
        STRING16  Size:16
        STRING8  Size:8
        STRING4  Size:4
        STRING2  Size:2
        N_PROF  Size:273
        N_PARAM  Size:5
        N_LEVELS  Size:496
        N_CALIB  Size:1
        N_HISTORY  Size:8   *** is unlimited ***

    9 global attributes:
        title: Argo float vertical profile
        institution: AOML
        source: Argo float
        history: 2016-01-04T20:15:15Z creation
        references: http://www.argodatamgt.org/Documentation
        comment: 
        user_manual_version: 3.03
        Conventions: Argo-3.0 CF-1.6
        featureType: trajectoryProfile
```

**Comment by richardsc** [comment]
That's a lot of stuff, but it looks to me like the only data contained in the NetCDF are: PRES, PSAL, TEMP, DOXY, and TEMP_DOXY (and all the associated QCs, adjustments, errors, etc), but none of the "other" fields I would have expected based on looking at the header from:

http://www.mbari.org/lobo/Data/FloatVizData/QC/8486HawaiiQC.txt

```
Cruise	
Station	
Type	
mon/day/yr	
hh:mm	
Lon [°E]	
Lat [°N]	
Bot. Depth[m]	
Depth[m]	QF	
Temperature[°C]	QF	
Salinity	QF	
Density	QF	
Oxygen[µM]	QF	
OxygenSat[%]	QF	
Nitrate[µM]	QF	
Chlorophyll[µg/l]	QF	
BackScatter[/m/sr]	QF	
pHinsitu[Total]	QF	
pH25C[Total]	QF
```

**Comment by dankelley** [comment]
I hacked together a little code that reads TAB-deliminated files. So far it ignores the QF lines, but that can be added without a lot of work. Below is the code, a few-line test file, and the results.

**R code**
```r
file <- "a"
d <- readLines(file)
header <- grepl("^//", d, useBytes=TRUE)
skip <- if (sum(header)) which(!header)[1] else 0
data <- read.delim(file, sep="\t", skip=skip)
names <- if (sum(header)) strsplit(d[skip], "\t")[[1]] else NULL
QF <- names == "QF"
Names <- names[!QF]
Data <- data[,!QF]
names(Data) <- Names
str(Data)

message("TO DO: assign each QF to column preceeding it, as in ODF")
```

**Test file**
```
//1																					
//Oxygen, 1.05
//Nitrate, 1, 1,0, 3.3
Cruise	Station	Type	mon/day/yr	hh:mm	Lon [°E]	Lat [°N]	Bot. Depth [m]	Depth[m]	QF	Temperature[°C]	QF	Salinity	QF	Density	QF	Oxygen[µM]	QF	OxygenSat[%]	QF	Nitrate[µM]	QF
5426DRAKEPASS	1	C	1/8/2009	2:08	280.6	-55.629		321.66	0	5.1354	0	34.1909	0	26.954	0	285.92	0	90.32	0	23.07	0
5426DRAKEPASS	1	C	1/8/2009	2:08	280.6	-55.629		311.38	0	5.1436	0	34.1883	0	26.951	0	286.02	0	90.37	0	22.75	0
5426DRAKEPASS	1	C	1/8/2009	2:08	280.6	-55.629		301.22	0	5.1385	0	34.185	0	26.949	0	286.55	0	90.52	0	22.57	0
```

**Results of running Rscript with this code and data**
```R
'data.frame':	2 obs. of  15 variables:
 $ Cruise         : Factor w/ 1 level "5426DRAKEPASS": 1 1
 $ Station        : int  1 1
 $ Type           : Factor w/ 1 level "C": 1 1
 $ mon/day/yr     : Factor w/ 1 level "1/8/2009": 1 1
 $ hh:mm          : Factor w/ 1 level "2:08": 1 1
 $ Lon [°E]       : num  281 281
 $ Lat [°N]       : num  -55.6 -55.6
 $ Bot. Depth [m] : logi  NA NA
 $ Depth[m]       : num  311 301
 $ Temperature[°C]: num  5.14 5.14
 $ Salinity       : num  34.2 34.2
 $ Density        : num  27 26.9
 $ Oxygen[µM]     : num  286 287
 $ OxygenSat[%]   : num  90.4 90.5
 $ Nitrate[µM]    : num  22.8 22.6
```

**Comment by dankelley** [comment]
@CaraWilson -- any chance you could check to see if the changes are OK for your issue? In oce, we want reporters to close issues. This one may have been fixed 8 days ago ... I don't quite know, because it is no longer "in my head". So, if things still need more work, please post something here to indicate clearly what is left to be done, and if not, please consider closing the issue. Thanks!  Dan.

**Comment by richardsc** [comment]
Hey @dankelley , I think our proposed plan of integrating reading those BioArgo files stalled when we started in on getting the units sorted out properly. Also, there were a couple other bugs discovered in the process related to `argo` objects (e.g. #816, #817, and #818).

As for an update for @CaraWilson, our plan is to provide some code, external to oce, that will read those files and convert them into `argo` objects. We decided for now that the MBARI file format is not stable enough to include officially in `oce`. The development work was started in a temporary repo which you can see at:

https://github.com/richardsc/mbari

but this is more of a sandbox right now. Sorry this is taking so long, but the good news is that we're squashing a bunch of other bugs in the process...

If you need a (less formal) solution sooner, you can adapt Dan's code above, or even the code in the sandbox repo I linked to above.

**Comment by CaraWilson** [comment]
While @richardsc provided me some code to read the bioargo file I had thought it was also going to be incorporated into the oce program.  The solution provided is fine with me, but I was not clear on what exactly was still being done as I realize it opened up a bunch of other issues that you both have been working on.  The reason I wanted to get the data into an oce data type was that I wanted to make a section plot of the data but I see that isn't yet possible as mentioned in #817. I was waiting to see a resolution on that to really play around with the data some more, but I will try out the mbar code provided and close out this issue if it seems to read the data in okay for me.  Does that sound okay?  I'm on an airplane most of the day, and then visiting my parents this weekend so I might not be able to do that until Tuesday.

**Comment by dankelley** [comment]
Hi. Since I'm the one assigned to this issue, I should make a decision! I don't think it is good for oce to be handling data formats that are temporary, so I think your best solution is to work with Clark on adjunct code.  It's easy for adjunct code to hook into oce (that is, to create objects that oce handles internally) so that will be your best path, I think.

I think this should be closed because the issue as posed in the title has been addressed.  (I am always on the lookout for issues to close because when there are a lot of issues, I can't see what I can tackle when I have a bit of free time.)

**Comment by CaraWilson** [comment]
I've finally been able to take a look at this and try to assess what @richardsc and @dankelley have been doing to see if I can now do what I want to do, which is use the oce package to make sections of biological parameters in the BioArgo data on the MBARI website. I am able to read the files into R, my hold-up was reading them into an oce dataset.  I initially thought that reading them into an ctd dataset would be the best way forward as that data type allowed the addition of parameters other than just temp and salinity, which the argo datatype did not.  Initially the issue was that the data was only available as depth, not pressure, but we found out from Ken Johnson that the parameter labelled "depth" in the MBARI files is really pressure (so named to not confuse the general public, and to stymie the scientists!), so that issue is mute.  I can read the barge into a ctd datatype using as.ctd and then ctdAddColumn, however I then realized that it seems that the section routines only use distance from first station for the x axis, whereas for float data one would want to use time instead.  Is this functionality available if they were read in as an argo datatype?  So I see that there are two issues: 
1) Add an argoAddColumn to append biological parameters to Argo datasets
2) Allow sections to be created using time (and latitude and longitude) as the x parameter.  

I really haven't used oce so I apologize if I has misinterpreted the functionality of it.

**Comment by richardsc** [comment]
Hi Cara,

You've basically got the idea (and this is something that is done quite commonly) -- to read the data in however necessary and then *coerce* into the class that you want.

I cleaned up the example repository [MBARI](https://github.com/richardsc/mbari), with an example that shows how to read the data into an `argo` object, which can then be converted to a `section` object for plotting. Note the use of `oceSetData()` in the `read_mbari_argo()` function to add the supplemental fields to the `argo` object (for your point 1 above).

The `plot` method for `section` objects has the following possibilities for the x-axis: `distance`, `track`, `longitude`, and `latitude`, and the example above shows how to make plots like that. Your request:

> 2) Allow sections to be created using time [...] as the x parameter.

is not currently possible, though I think it makes sense to add that functionality. We'll look into it.

However, in creating the above example, I discovered that trying to plot either `oxygenSaturation` or `nitrate` in the section fails for reasons I don't understand yet. Plotting `which='oxygen'` works. Will look at this tomorrow.

**Comment by CaraWilson** [comment]
Thanks. I downloaded the code from the MBARI repository, however it didn't work with my file of interest,  http://www.mbari.org/lobo/Data/FloatVizData/QC/8486HawaiiQC.txt. This file has 4 more variables after nitrate: chlorophyll, Backscatter, pHinsitu and ph25C, but none of these were read both the code.  

**Result of running the code:** 
 ```r
d <- read.argo.mbari("8486HawaiiQC.txt")
There were 14 warnings (use warnings() to see them)
> summary(d) # print a summary of the object
Argo Summary
------------

* source:     "/Users/CaraWilson/Data/BioArgo/8486HawaiiQC.txt"
* id:         "NA"
* Profiles:   "0 delayed; 0 adjusted; 0 realtime"
* Time ranges from 2013-05-20 14:48:00 to 2015-12-17 00:47:00 with mean increment 5.028962 days
* Statistics of data
                                   Min.       Mean      Max.
     longitude [°E]           -158.3120 -151.74248 -147.4120
     latitude [°N]              22.2310   24.19709   26.2390
     salinity [PSS-78]          33.9782   34.69640   35.5728
     temperature [°C, ITS-90]    2.7368   14.86058   27.0552
     pressure [dbar]             4.3600  355.94238 1500.8900
     oxygen [μM]                13.9100  163.62733  239.0600
     oxygenSaturation            4.3400   68.06764  109.4100
     nitrate [μM]               -1.5300   14.72586   46.4400
* Processing Log::
  * 2016-01-21 11:30:53 UTC: ``create 'argo' object``
  * 2016-01-21 11:30:53 UTC: ``as.argo(time = time, longitude = ifelse(longitude > 180, longitude -     360, longitude), latitude = latitude, salinity = salinity,     temperature = temperature, pressure = pressure, filename = filename)``
  * 2016-01-21 11:30:53 UTC: ``oceSetData(object = d, name = "oxygen", value = oxygen, units = list(unit = expression(mu *     M), scale = ""))``
  * 2016-01-21 11:30:53 UTC: ``oceSetData(object = d, name = "oxygenSaturation", value = oxygenSaturation,     units = list(unit = expression(), scale = ""))``
  * 2016-01-21 11:30:53 UTC: ``oceSetData(object = d, name = "nitrate", value = nitrate, units = list(unit = expression(mu *     M), scale = ""))``
  * 2016-01-21 11:30:53 UTC: ``oceSetMetadata(object = d, name = "flags", value = list(pressure = pressureFlag,     temperature = temperatureFlag, salinity = salinityFlag, oxygen = oxygenFlag,     oxygenSaturation = oxygenSaturationFlag, nitrate = nitrateFlag))``
```

**Comment by richardsc** [comment]
Ah, I see. The file at:

http://www3.mbari.org/lobo/data/floatvizdata/floatvizconfig.txt

lists all the possible fields, which obviously vary from file to file. A more general form of the function should recognize those too. I'll update the `mbari` repo when I get a chance. I'll also see what those warnings are.
```
Nitrate[µM]
Depth[m]
mon/day/yr
Salinity
Temperature[°C]
Density
Oxygen[µM]
OxygenSat[%]
Chlorophyll[µg/l]
BackScatter[/m/sr]
CDOM[PPB]
pHinsitu[Total]
pH25C[Total]
Lon [°E]
Lat [°N]
```

**Comment by dankelley** [comment]
I'm having a look at having `read.argo()` read the extra columns in the standardized netcdf files. I don't think it's going to be terribly hard but I will need to sort out some wrinkles. For example, the Dropbox file named in a previous comment (https://github.com/dankelley/oce/issues/812#issuecomment-168795128) states that `DOXY` is in umol/kg, but at the moment I am using `ODFNames2oceNames()` to create a variable name, and this yields `oxygen_by_volume`, which is obviously not the same.  I suspect that `ODFNames2oceNames()` is simply wrong ... but in the docs for that I do not state the URL for the correspondence (bad me) so I'll have to do some digging. Possibly I should just copy `ODFNames2oceNames()` to `ArgoNames2oceNames()` and tailor it, since it may be too much to expect different agencies to agree exactly on variable names.

I'll look at this on the weekend.

**Comment by dankelley** [comment]
Hm. BIO has a site listing variable codes (http://www.isdm.gc.ca/isdm-gdsi/diction/code_search-eng.asp?code=DOXY) and below is a screenshot. So the name alone is not going to be enough for `read.argo()`; it will have to look in more detail.
![screen shot 2016-01-22 at 11 44 31 am](https://cloud.githubusercontent.com/assets/99469/12515217/a7f28094-c0fd-11e5-80ed-c292830e5831.png)

**Comment by richardsc** [comment]
Hi @CaraWilson . Can you check out the example at the [mbari](https://github.com/richardsc/mbari) repo now? I think I've got it working more generally, and specifically for your example file. You'll note there are still some lingering argo and section bugs (#829, #833, and #836), but you should be able to access and plot the data as either an `argo` or `section` object as needed.

**Comment by CaraWilson** [comment]
Wah hoo - that is friggin' awesome!  Yes, it worked like a charm.  IMO this issue can be closed now.  Thanks so much to both of you @richardsc  and @dankelley for all the work you put into this. It turned into something much bigger than in initially seemed.  I owe you both a beer at Ocean Sciences! Unfortunately I can't play around with it this weekend - looking forward to Monday when I will be able to!!! 
Thanks again.

**Comment by dankelley** [comment]
Closing, according to the last comment from Cara. It can always be reopened, of course.

**Comment by CaraWilson** [comment]
Hi @richardsc,  I finally got a chance to play around more with the BioArgo data using your new code, and I discovered that the code works for some files, but doesn’t work for a lot of them, producing the following error: 
```r
> float <- read.argo.mbari("7622HawaiiQC.txt")
Error in res[i, 1:n] <- rev(x[[i]][[field]]) : 
  number of items to replace is not a multiple of replacement length
```

Note that MBARI has renamed their websites and the bioArgo files are now on a completely different website: 

http://www.mbari.org/science/upper-ocean-systems/chemical-sensor-group/floatviz/

**Comment by dankelley** [comment]
Is it possible to supply a URL to the file in question?  It's hard to tell what is going on, without sample data. If the website doesn't provide clear URLs (a bad sign, in my view ...) then maybe you could email Clark and me a copy of the file?

**Comment by CaraWilson** [comment]
http://www3.mbari.org/lobo/Data/FloatVizData/QC/7622HawaiiQC.txt

**Comment by dankelley** [comment]
Thanks. I see why this is happening. It is because the file lacks a field (chlorophyll) that was 'hard-coded' in Clark's solution. I normally solve this by wrapping a lot of text in statements like
```r
if ("chlorophyll" %in% names) {
  # do something
}
```
but there are more clever ways with function closures. In any case, Clark is likely to be able to see the solution quicker than I might.  I know he is a clever person because he was visiting Dalhousie today and gave what several 2nd year grad students said was the best talk they had seen.  (Trying to make Clark blush here ... no worries, since he is likely out networking right now.)

PS. Cara, in case you're interested, this is the way to find out where problems arise, so sometimes you can make a quick hack to code when an author cannot be reached quickly...

First, re-source (useful for library code because otherwise line numbers will be guesses) and re-run your test.
```r
source('read_mbari_argo.R')
d<-read.argo.mbari('~/7622HawaiiQC.txt')
```
we see
```r
Error in res[i, 1:n] <- rev(x[[i]][[field]]) : 
  number of items to replace is not a multiple of replacement length
```
and we can find where the problem is with
```r
traceback()
```
which yields 
2: makeMatrix(D, "chlorophyll") at read_mbari_argo.R#110
1: read.argo.mbari("~/7622HawaiiQC.txt")
```
and then we can pop into an editor to inspect the code, viz.
<img width="447" alt="screen shot 2016-01-26 at 7 15 28 pm" src="https://cloud.githubusercontent.com/assets/99469/12598526/35c14b38-c461-11e5-9605-cc3a27dc0127.png">

**Comment by CaraWilson** [comment]
Okay, thanks for the detailed explanation.  

On an Argo related note, we also have all of the regular argo data on our ERDDAP. Unfortunately the server we are getting the data from doesn’t retain any of the ARGO id numbers, so there is no way to know which float a profile is from, or to make transects of one float. However if you are interested in getting all profiles from a certain box for a certain timeperiod, which I have wanted to do for some past projects, its much easier getting the data from ERDDAP than sorting through all the different directories on the main ARGO data distribution.

**Comment by dankelley** [comment]
Cara, can the erddap server provide (pressure, salinity, temperature)? I couldn't figure out how to indicate that in the webform. I also tried various things in constructing a URL.

Also, regarding the IDs, will that be coming eventually?

On the 'news' front, I think on Saturday I'll write a new oce function that will do queries to erddap (and, perhaps, other servers). Below is a code that uses a query that I got from the webserver (after removing plotting things) and a graph made up by it. When I have the function done, users won't have to figure out these queries themselves, and if the query changes, then it's a simple matter of updating oce.

```r
## 01_argo.R: Download and plot some argo traces.
## Q: how to get (pressure,temperature,salinity) data?

## Cache data in 01_argo.nc locally.
if (0 == length(list.files(pattern="01_argo.nc"))) {
    download.file('http://coastwatch.pfeg.noaa.gov/erddap/tabledap/apdrcArgoAll.nc?longitude,latitude,time&time%3E=2005-01-21T00:00:00Z&time%3C=2011-01-28T00:00:00Z&longitude%3E=277.3&longitude%3C=350.1&latitude%3E=45&latitude%3C=90', "01_argo.nc")
}

library(ncdf4)
f <- nc_open("01_argo.nc")
lon <- ncvar_get(f, "longitude")
lat <- ncvar_get(f, "latitude")

library(oce)
data(coastlineWorldMedium, package="ocedata")
if (!interactive()) png("01_argo.png")
par(mar=c(2, 2, 1, 1))
mapPlot(coastlineWorldMedium, longitudelim=c(-60, -10), latitudelim=c(45,75),
        fill='lightgray', projection='+proj=lcc +lon_0=-35')
mapPoints(lon, lat)

if (!interactive()) dev.off()
```

![01_argo](https://cloud.githubusercontent.com/assets/99469/12615265/0c8c52d8-c4db-11e5-8d78-8f482d220066.png)

**Comment by CaraWilson** [comment]
Yes, ERDDAP serves the temp, salinity, pressure data for the argo data.  From the graph page: 

http://coastwatch.pfeg.noaa.gov/erddap/tabledap/apdrcArgoAll.graph

click the "Data Access Link" in the upper right, or just replace the "graph" in the url above to "html": 

http://coastwatch.pfeg.noaa.gov/erddap/tabledap/apdrcArgoAll.html 

Either method will bring up a form where you can pick which parameters you want, and also place constraints on them if desired.  Once that is done you can see what url will be generated by that request.

**Comment by CaraWilson** [comment]
As far as I know, there are no plans to serve the argo identifier code

**Comment by richardsc** [comment]
Hi @CaraWilson,

@dankelley was right about why the code failed for that file -- I coded it quickly to handle the one you had, and made it brittle to input file differences. Dan also provided a good partial solution with the `try()` block.

I'll update later today or this weekend, when I get a chance, but if you're interested you might want to try doing it yourself first. The best way to learn these things is "on the battlefield" so to speak! :)

**Comment by CaraWilson** [comment]
Thanks @richardsc.  I'm stuck in a meeting today and tomorrow, so I won't have a chance to test it out myself in the next few days.

**Comment by dankelley** [comment]
I am trying to clear up issues in preparation for a CRAN release, This issue is not marked `Next Release`, so it will not block the process to a CRAN release, but I like to have a clear(ish) issue screen when I go into oce to find a task to fill some time, so I would like to encourage @CaraWilson to consider closing the issue, since

1. We have decided not to include support for this temporary file format in oce.  We prefer to focus our development efforts on more permanent file formats, especially ones that with published format specifications.

2. A working non-oce solution is just a step or two away, and discussion of the remaining steps can occur just as easily in a closed issue as in an open issue.

3. The issue can always be reopened if it returns topics that fit more naturally into the oce category.
